### PR TITLE
Get started in contquadratic tests

### PR DESCRIPTION
**Pull request by joaquimg**

### CODE REVIEW

**Comment by joaquimg** [comment]
Start porting quadratic tests detailed in #23

File: test/contquadratic.jl [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            v = MOI.addvariables!(m, 3)
+            @test MOI.getattribute(m, MOI.VariableCount()) == 3
+
+            c1 = MOI.addconstraint!(m, MOI.ScalarAffineFunction(v, [1.0,2.0,3.0], 0.0), MOI.GreaterThan(4.0))
+            @test MOI.getattribute(m, MOI.ConstraintCount()) == 1
+
+            c2 = MOI.addconstraint!(m, MOI.ScalarAffineFunction([v[1],v[2]], [1.0,1.0], 0.0), MOI.GreaterThan(1.0))
+            @test MOI.getattribute(m, MOI.ConstraintCount()) == 2
+
+            MOI.setobjective!(m, MOI.MinSense, MOI.ScalarQuadraticFunction(v, [0.0,0.0,0.0],[v[1], v[1], v[2], v[2], v[3]], [v[1], v[2], v[2], v[3], v[3]], [2.0, 1.0, 2.0, 1.0, 2.0], 0.0))
```
Comment (mlubin) [inline test/contquadratic.jl]: Are the off-diagonals scaled correctly? Shouldn't it be ``[2.0, 0.5, 2.0, 0.5, 2.0]``? These are coefficients in Q and the objective is (1/2).

File: test/contquadratic.jl [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            @test MOI.getattribute(m, MOI.Sense()) == MOI.MinSense
+
+            if MOI.cangetattribute(m, MOI.ObjectiveFunction())
+                obj = MPI.getattribute(m, MOI.ObjectiveFunction())
```
Comment (mlubin) [inline test/contquadratic.jl]: We need to document the structure we expect for returned functions. It's probably not reasonable to expect the quadratic function to be returned in the exact same order as was given for input. We could possibly impose CSC format with increasing indices within a column, or don't impose anything and make comparisons after normalizing.

File: test/contquadratic.jl [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                @test obj.affine_variables == v
+                @test obj.affine_coefficients == [0.0,0.0,0.0]
+                @test obj.quadratic_rowvariables == [v[1], v[1], v[2], v[2], v[3]]
+                @test obj.quadratic_colvariables == [v[1], v[2], v[2], v[3], v[3]]
+                @test obj.quadratic_coefficients == [2.0, 1.0, 2.0, 1.0, 2.0]
+                @test obj.constant == 0.0
+            end
+
+            if MOI.cangetattribute(m, MOI.ConstraintFunction(), c1)
+                aff = MOI.getattribute(m, MOI.ConstraintFunction(), c1)
```
Comment (mlubin) [inline test/contquadratic.jl]: We probably need a similar discussion here. Are affine functions returned in the exact same order that you provided them? I didn't think about this when I was writing some of the linear tests.

File: test/contquadratic.jl (hunk 1 above)
Comment (mlubin) [inline test/contquadratic.jl]: See https://github.com/JuliaOpt/MathProgBase.jl/blob/bdaaa744d0ef43e95dfd6b2f3eea9c8fbd6e2a9d/test/quadprog.jl#L19

File: test/contquadratic.jl [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            v = MOI.addvariables!(m, 3)
+            @test MOI.getattribute(m, MOI.VariableCount()) == 3
+
+            c1 = MOI.addconstraint!(m, MOI.ScalarAffineFunction(v, [1.0,2.0,3.0], 0.0), MOI.GreaterThan(4.0))
+            @test MOI.getattribute(m, MOI.ConstraintCount()) == 1
+
+            c2 = MOI.addconstraint!(m, MOI.ScalarAffineFunction([v[1],v[2]], [1.0,1.0], 0.0), MOI.GreaterThan(1.0))
+            @test MOI.getattribute(m, MOI.ConstraintCount()) == 2
+
+            MOI.setobjective!(m, MOI.MinSense, MOI.ScalarQuadraticFunction(v, [0.0,0.0,0.0],[v[1], v[1], v[1], v[2], v[2], v[3], v[3]], [v[1], v[2], v[2], v[2], v[3], v[3], v[3]], [2.0, 0.5, 0.5, 2.0, 1.0, 1.0, 1.0], 0.0))
```
Comment (mlubin) [inline test/contquadratic.jl]: Same comment about the scaling of the off-diagonal terms

File: test/contquadratic.jl [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+            @test MOI.supportsproblem(solver, MOI.ScalarAffineFunction, [(MOI.VectorAffineFunction{Float64},MOI.NonPositive),(MOI.ScalarQuadraticFunction{Float64},MOI.LessThan)])
+
+            m = MOI.SolverInstance(solver)
+
+            x = MOI.addvariable!(m)
+            y = MOI.addvariable!(m)
+            @test MOI.getattribute(m, MOI.VariableCount()) == 2
+
+            c1 = MOI.addconstraint!(m, MOI.VectorAffineFunction([1,2], [x,y,x,y],[-1.0,1.0,1.0,1.0], [0.0,0.0]), MOI.NonPositive(2))
```
Comment (mlubin) [inline test/contquadratic.jl]: Shouldn't this be ``NonNegative``?

File: test/contquadratic.jl (hunk 2 above)
Comment (joaquimg) [inline test/contquadratic.jl]: I like CSC format with increasing indices.
But we are returning Variable references right?
Are we writing our own Vector of triplets(or triple of vectors)?

File: test/contquadratic.jl (hunk 3 above)
Comment (joaquimg) [inline test/contquadratic.jl]: I think affine functions should be returned in increasing order also...

File: test/contquadratic.jl (hunk 5 above)
Comment (joaquimg) [inline test/contquadratic.jl]: Yes, `NonNegative`

File: test/contquadratic.jl (hunk 1 above)
Comment (joaquimg) [inline test/contquadratic.jl]: On this example, we are ignoring duplicates or adding them?

Comment (mlubin) [inline test/contquadratic.jl]: Which are the duplicates here? Specifying (i,j,v) also specifies (j,i,v).

File: test/contquadratic.jl (hunk 2 above)
Comment (mlubin) [inline test/contquadratic.jl]: There's no real ordering on the variable references. Let me throw something together.

File: test/contquadratic.jl (hunk 1 above)
Comment (joaquimg) [inline test/contquadratic.jl]: the example you posted is:
```julia

rows = [1, 2, 2, 2, 3, 3, 3]
cols = [1, 1, 1, 2, 2, 3, 3]
vals = [2.0, 0.5, 0.5, 2.0, 1.0, 1.0, 1.0]
```
Here `(2,1,0.5)` appears twice. Same for `(3,3,1.0)`. They are being added right?

The triplets I wrote are:
```julia
rows = [1,1,2,2,3]
cols = [1,2,2,3,3]
vals = [2.0, 1.0, 2.0, 1.0, 2.0]
```

Comment (mlubin) [inline test/contquadratic.jl]: Good point. So is the objective in the comment ``x^2 + 0.5xy + y^2 + 0.5yz + z^2`` correct?

Comment (joaquimg) [inline test/contquadratic.jl]: Lest see, my understanding is that the values:
```julia
rows = [1,1,2,2,3]
cols = [1,2,2,3,3]
vals = [2.0, 1.0, 2.0, 1.0, 2.0]
```
imply in the Matrix:
```julia
Q = [[2.0, 1.0, 0.0],
     [1.0, 2.0, 1.0],
     [0.0, 1.0, 2.0]]
```
And thus the objective should be: `x^2 + xy + y^2 + yz + z^2`, right?

Comment (mlubin) [inline test/contquadratic.jl]: Yes, that seems right

File: test/contquadratic.jl [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+            @test MOI.supportsproblem(solver, MOI.ScalarAffineFunction, [(MOI.VectorAffineFunction{Float64},MOI.NonNegative),(MOI.ScalarQuadraticFunction{Float64},MOI.LessThan)])
+
+            m = MOI.SolverInstance(solver)
+
+            x = MOI.addvariable!(m)
+            y = MOI.addvariable!(m)
+            @test MOI.getattribute(m, MOI.VariableCount()) == 2
+
+            c1 = MOI.addconstraint!(m, MOI.VectorAffineFunction([1,2], [x,y,x,y],[-1.0,1.0,1.0,1.0], [0.0,0.0]), MOI.NonNegative(2))
```
Comment (mlubin) [inline test/contquadratic.jl]: The set is now called `Nonnegative`

**Comment by mlubin** [comment]
We haven't yet defined our duality conventions for QPs, so these tests need some discussion.

**Comment by joaquimg** [comment]
Should I comment the getting duals part? So we can merge.
Or its better to just wait for the discussion?
I could also get the manual on QCQP duals started here by following Boyd's Convex Optimization...

**Comment by mlubin** [comment]
Yes, you can comment out the duals for now

File: test/contquadratic.jl [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+function contquadratictest(solver::MOI.AbstractSolver, ε=Base.rtoldefault(Float64))
+    @testset "Testing quadratic interface with $solver" begin
+        @testset "QP1 - Quadratic objective" begin
+            # simple quadratic objective
+            # Min x^2 + xy + y^2 + yz + z^2
+            # st  x + 2y + 3z >= 4 (c1)
+            #     x +  y      >= 1 (c2)
+            #     x,y \in R
+
+            @test MOI.supportsproblem(solver, MOI.ScalarQuadraticFunction, [(MOI.ScalarAffineFunction{Float64},MOI.GreaterThan)])
```
Comment (mlubin) [inline test/contquadratic.jl]: Missing type parameter on ``MOI.ScalarQuadraticFunction``

File: test/contquadratic.jl (hunk 3 above)
Comment (mlubin) [inline test/contquadratic.jl]: This should now be updated to ``aff \approx ...``

File: test/contquadratic.jl (hunk 2 above)
Comment (mlubin) [inline test/contquadratic.jl]: This should now be replaced with an `isapprox` call (and we also need an implementation of Base.isapprox for two scalar quadratic functions)

File: test/contquadratic.jl [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        @testset "QP1" begin
+            # same as QP0 but with duplicate terms
+            # then change the objective and sense
+            # simple quadratic objective
+            # Min x^2 + xy + y^2 + yz + z^2
+            # st  x + 2y + 3z >= 4 (c1)
+            #     x +  y      >= 1 (c2)
+            #     x,y \in R
+
+            @test MOI.supportsproblem(solver, MOI.ScalarQuadraticFunction, [(MOI.ScalarAffineFunction{Float64},MOI.GreaterThan)])
```
Comment (mlubin) [inline test/contquadratic.jl]: Missing type parameter on MOI.ScalarQuadraticFunction

File: test/contquadratic.jl [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+            c2 = MOI.addconstraint!(m, MOI.ScalarAffineFunction([v[1],v[2]], [1.0,1.0], 0.0), MOI.GreaterThan(1.0))
+            @test MOI.getattribute(m, MOI.ConstraintCount()) == 2
+
+            obj = MOI.ScalarQuadraticFunction(v, [0.0,0.0,0.0],[v[1], v[1], v[1], v[2], v[2], v[3], v[3]], [v[1], v[2], v[2], v[2], v[3], v[3], v[3]], [2.0, 0.5, 0.5, 2.0, 1.0, 1.0, 1.0], 0.0)
+            MOI.setobjective!(m, MOI.MinSense, obj)
+            @test MOI.getattribute(m, MOI.Sense()) == MOI.MinSense
+
+            if MOI.cangetattribute(m, MOI.ObjectiveFunction())
+                @test obj ≈ MPI.getattribute(m, MOI.ObjectiveFunction())
```
Comment (mlubin) [inline test/contquadratic.jl]: This isn't defined yet

File: test/contquadratic.jl (hunk 8 above)
Comment (joaquimg) [inline test/contquadratic.jl]: This type is missing in linear tests also!

File: test/contquadratic.jl (hunk 9 above)
Comment (joaquimg) [inline test/contquadratic.jl]: Should it be defined? Or do I remove from the test?

File: test/contquadratic.jl (hunk 8 above)
Comment (mlubin) [inline test/contquadratic.jl]: My bad

File: test/contquadratic.jl (hunk 9 above)
Comment (mlubin) [inline test/contquadratic.jl]: It should be defined (you should define it)

Comment (joaquimg) [inline test/contquadratic.jl]: ok!

File: test/contquadratic.jl [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+            # st  x + 2y + 3z >= 4 (c1)
+            #     x +  y      >= 1 (c2)
+            #     x,y \in R
+
+            @test MOI.supportsproblem(solver, MOI.ScalarQuadraticFunction{Float64}, [(MOI.ScalarAffineFunction{Float64},MOI.GreaterThan)])
+
+            m = MOI.SolverInstance(solver)
+
+            v = MOI.addvariables!(m, 3)
+            @test MOI.getattribute(m, MOI.VariableCount()) == 3
```
Comment (mlubin) [inline test/contquadratic.jl]: ``VariableCount`` was renamed to ``NumberOfVariables``

File: test/contquadratic.jl [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+
+            c1 = MOI.addconstraint!(m, MOI.ScalarAffineFunction(v, [1.0,2.0,3.0], 0.0), MOI.GreaterThan(4.0))
+            @test MOI.getattribute(m, MOI.ConstraintCount()) == 1
```
Comment (mlubin) [inline test/contquadratic.jl]: ``ConstraintCount`` was renamed to ``NumberOfConstraints{F,S}``, now specific to the type of constraint

Comment (mlubin) [inline test/contquadratic.jl]: See the linear tests

**Comment by mlubin** [comment]
Looks good! There are just two places now at the beginning of the tests which still have the old code for comparisons instead of \approx.

**Pull request merged**